File: src/lib/video-cache.ts
import fs from 'fs';
import path from 'path';

const CACHE_DIR = path.join('E:', '社院课程stt', 'video_cache');
const MAX_CACHE_SIZE = 20 * 1024 * 1024 * 1024;
const META_FILE = path.join(CACHE_DIR, 'cache_meta.json');

interface CacheEntry {
  size: number;
  lastAccess: number;
}

let initialized = false;
const cacheMeta = new Map<string, CacheEntry>();
let evictionInProgress = false;

function ensureCacheDir(): void {
  if (!fs.existsSync(CACHE_DIR)) {
    fs.mkdirSync(CACHE_DIR, { recursive: true });
  }
}

function loadMeta(): void {
  try {
    if (fs.existsSync(META_FILE)) {
      const data = JSON.parse(fs.readFileSync(META_FILE, 'utf-8'));
      for (const [key, value] of Object.entries(data)) {
        cacheMeta.set(key, value as CacheEntry);
      }
    }
  } catch {
    // 元数据损坏，重新扫描
    cacheMeta.clear();
  }
}

function saveMetaSync(): void {
  try {
    ensureCacheDir();
    const obj: Record<string, CacheEntry> = {};
    for (const [key, value] of cacheMeta.entries()) {
      obj[key] = value;
    }
    fs.writeFileSync(META_FILE, JSON.stringify(obj), 'utf-8');
  } catch {
    // 保存失败不影响缓存功能
  }
}

function saveMetaAsync(): void {
  setImmediate(() => saveMetaSync());
}

function getTotalCacheSize(): number {
  let total = 0;
  for (const entry of cacheMeta.values()) {
    total += entry.size;
  }
  return total;
}

function evictLRU(neededSpace: number): void {
  if (evictionInProgress) return;
  evictionInProgress = true;

  try {
    const entries = Array.from(cacheMeta.entries())
      .sort((a, b) => a[1].lastAccess - b[1].lastAccess);

    let freed = 0;
    for (const [code, entry] of entries) {
      if (getTotalCacheSize() - freed + neededSpace <= MAX_CACHE_SIZE) {
        break;
      }

      const cachePath = path.join(CACHE_DIR, `${code}.mp4`);
      try {
        if (fs.existsSync(cachePath)) {
          fs.unlinkSync(cachePath);
        }
        freed += entry.size;
        cacheMeta.delete(code);
        console.log(`[视频缓存] LRU淘汰: ${code} (${(entry.size / 1024 / 1024).toFixed(1)}MB)`);
      } catch {
        cacheMeta.delete(code);
      }
    }

    if (freed > 0) {
      saveMetaAsync();
    }
  } finally {
    evictionInProgress = false;
  }
}

export function initVideoCache(): void {
  if (initialized) return;

  ensureCacheDir();
  loadMeta();

  // 清理元数据中不存在的文件
  let cleaned = false;
  for (const [code] of cacheMeta.entries()) {
    const cachePath = path.join(CACHE_DIR, `${code}.mp4`);
    if (!fs.existsSync(cachePath)) {
      cacheMeta.delete(code);
      cleaned = true;
    }
  }

  // 扫描缓存目录中未被元数据记录的文件
  try {
    const files = fs.readdirSync(CACHE_DIR);
    for (const file of files) {
      if (!file.endsWith('.mp4')) continue;
      const code = file.replace('.mp4', '');
      if (!cacheMeta.has(code)) {
        const cachePath = path.join(CACHE_DIR, file);
        try {
          const stat = fs.statSync(cachePath);
          cacheMeta.set(code, {
            size: stat.size,
            lastAccess: stat.mtimeMs,
          });
          cleaned = true;
        } catch {
          // 跳过无法读取的文件
        }
      }
    }
  } catch {
    // 扫描失败不影响
  }

  if (cleaned) {
    saveMetaAsync();
  }

  const totalMB = getTotalCacheSize() / (1024 * 1024);
  console.log(`[视频缓存] 初始化完成: ${cacheMeta.size} 个文件, ${totalMB.toFixed(0)}MB / ${(MAX_CACHE_SIZE / 1024 / 1024).toFixed(0)}MB`);

  initialized = true;
}

export function getCachePath(courseCode: string): string | null {
  initVideoCache();

  const cachePath = path.join(CACHE_DIR, `${courseCode}.mp4`);
  if (fs.existsSync(cachePath)) {
    const now = Date.now();
    const entry = cacheMeta.get(courseCode);
    if (entry) {
      entry.lastAccess = now;
    } else {
      try {
        const stat = fs.statSync(cachePath);
        cacheMeta.set(courseCode, {
          size: stat.size,
          lastAccess: now,
        });
      } catch {
        return null;
      }
    }

    // 更新时间戳（降级到可读模式，每60秒写一次）
    if (now - lastMetaSave > 60000) {
      lastMetaSave = now;
      saveMetaAsync();
    }
    return cachePath;
  }

  // 文件被意外删除，清理元数据
  cacheMeta.delete(courseCode);
  return null;
}

let lastMetaSave = 0;

export function cacheVideo(courseCode: string, nasPath: string): void {
  initVideoCache();

  const cachePath = path.join(CACHE_DIR, `${courseCode}.mp4`);

  // 如果已在缓存中，只更新时间
  if (fs.existsSync(cachePath)) {
    const entry = cacheMeta.get(courseCode);
    if (entry) {
      entry.lastAccess = Date.now();
    }
    return;
  }

  // 后台异步拷贝
  setImmediate(() => {
    try {
      const nasStat = fs.statSync(nasPath);
      const fileSize = nasStat.size;

      // 先淘汰旧文件腾空间
      evictLRU(fileSize);

      // 拷贝文件
      ensureCacheDir();
      const tmpPath = cachePath + '.tmp';
      fs.copyFileSync(nasPath, tmpPath);
      fs.renameSync(tmpPath, cachePath);

      cacheMeta.set(courseCode, {
        size: fileSize,
        lastAccess: Date.now(),
      });

      saveMetaAsync();
      console.log(`[视频缓存] 已缓存: ${courseCode} (${(fileSize / 1024 / 1024).toFixed(1)}MB)`);
    } catch (err) {
      console.error(`[视频缓存] 拷贝失败 ${courseCode}:`, err);
      // 清理临时文件
      const tmpPath = cachePath + '.tmp';
      try {
        if (fs.existsSync(tmpPath)) fs.unlinkSync(tmpPath);
      } catch {
        // 忽略
      }
    }
  });
}

export function getCacheStats(): { count: number; totalSize: number; maxSize: number } {
  initVideoCache();
  return {
    count: cacheMeta.size,
    totalSize: getTotalCacheSize(),
    maxSize: MAX_CACHE_SIZE,
  };
}

File: knowledge-server/video_cache.py
import os
import json
import shutil
import threading
import time
from pathlib import Path
from typing import Optional

CACHE_DIR = Path(r"E:\社院课程stt\video_cache")
MAX_CACHE_SIZE = 20 * 1024 * 1024 * 1024
META_FILE = CACHE_DIR / "cache_meta.json"

_cache_meta: dict[str, dict] = {}
_lock = threading.Lock()
_initialized = False
_eviction_in_progress = False
_last_meta_save = 0


def _ensure_cache_dir():
    CACHE_DIR.mkdir(parents=True, exist_ok=True)


def _load_meta():
    global _cache_meta
    try:
        if META_FILE.exists():
            with open(META_FILE, "r", encoding="utf-8") as f:
                _cache_meta = json.load(f)
    except Exception:
        _cache_meta = {}


def _save_meta():
    try:
        _ensure_cache_dir()
        with open(META_FILE, "w", encoding="utf-8") as f:
            json.dump(_cache_meta, f, ensure_ascii=False)
    except Exception:
        pass


def _get_total_size() -> int:
    return sum(entry.get("size", 0) for entry in _cache_meta.values())


def _evict_lru(needed_space: int):
    global _eviction_in_progress
    if _eviction_in_progress:
        return
    _eviction_in_progress = True

    try:
        entries = sorted(
            _cache_meta.items(),
            key=lambda x: x[1].get("lastAccess", 0),
        )

        freed = 0
        for code, entry in entries:
            if _get_total_size() - freed + needed_space <= MAX_CACHE_SIZE:
                break

            cache_path = CACHE_DIR / f"{code}.mp4"
            try:
                if cache_path.exists():
                    cache_path.unlink()
                freed += entry.get("size", 0)
                del _cache_meta[code]
                size_mb = entry.get("size", 0) / (1024 * 1024)
                print(f"[视频缓存] LRU淘汰: {code} ({size_mb:.1f}MB)")
            except Exception:
                if code in _cache_meta:
                    del _cache_meta[code]

        if freed > 0:
            _save_meta()
    finally:
        _eviction_in_progress = False


def init_video_cache():
    global _initialized
    if _initialized:
        return

    with _lock:
        if _initialized:
            return

        _ensure_cache_dir()
        _load_meta()

        cleaned = False

        for code in list(_cache_meta.keys()):
            cache_path = CACHE_DIR / f"{code}.mp4"
            if not cache_path.exists():
                del _cache_meta[code]
                cleaned = True

        try:
            for f in CACHE_DIR.iterdir():
                if not f.name.endswith(".mp4"):
                    continue
                code = f.name.replace(".mp4", "")
                if code not in _cache_meta:
                    stat = f.stat()
                    _cache_meta[code] = {
                        "size": stat.st_size,
                        "lastAccess": stat.st_mtime * 1000,
                    }
                    cleaned = True
        except Exception:
            pass

        if cleaned:
            _save_meta()

        total_mb = _get_total_size() / (1024 * 1024)
        max_mb = MAX_CACHE_SIZE / (1024 * 1024)
        print(f"[视频缓存] 初始化完成: {len(_cache_meta)} 个文件, {total_mb:.0f}MB / {max_mb:.0f}MB")

        _initialized = True


def get_cache_path(course_code: str) -> Optional[str]:
    global _last_meta_save
    init_video_cache()

    cache_path = CACHE_DIR / f"{course_code}.mp4"

    with _lock:
        if cache_path.exists():
            now = time.time() * 1000
            if course_code in _cache_meta:
                _cache_meta[course_code]["lastAccess"] = now
            else:
                try:
                    stat = cache_path.stat()
                    _cache_meta[course_code] = {
                        "size": stat.st_size,
                        "lastAccess": now,
                    }
                except Exception:
                    return None

            if now - _last_meta_save > 60000:
                _last_meta_save = now
                _save_meta()

            return str(cache_path)

        if course_code in _cache_meta:
            del _cache_meta[course_code]

    return None


def cache_video(course_code: str, nas_path: str):
    init_video_cache()

    cache_path = CACHE_DIR / f"{course_code}.mp4"

    with _lock:
        if cache_path.exists():
            if course_code in _cache_meta:
                _cache_meta[course_code]["lastAccess"] = time.time() * 1000
            return

    def _copy():
        try:
            nas = Path(nas_path)
            file_size = nas.stat().st_size

            with _lock:
                _evict_lru(file_size)

            _ensure_cache_dir()
            tmp_path = CACHE_DIR / f"{course_code}.mp4.tmp"
            shutil.copy2(str(nas), str(tmp_path))
            tmp_path.replace(cache_path)

            with _lock:
                _cache_meta[course_code] = {
                    "size": file_size,
                    "lastAccess": time.time() * 1000,
                }
                _save_meta()

            print(f"[视频缓存] 已缓存: {course_code} ({file_size / (1024 * 1024):.1f}MB)")
        except Exception as e:
            print(f"[视频缓存] 拷贝失败 {course_code}: {e}")
            tmp_path = CACHE_DIR / f"{course_code}.mp4.tmp"
            try:
                if tmp_path.exists():
                    tmp_path.unlink()
            except Exception:
                pass

    threading.Thread(target=_copy, daemon=True).start()


def get_cache_stats() -> dict:
    init_video_cache()
    with _lock:
        return {
            "count": len(_cache_meta),
            "totalSize": _get_total_size(),
            "maxSize": MAX_CACHE_SIZE,
        }
